### ISSUE TITLE
YouTube Integration

### ISSUE THREAD

**Issue by z-huang** [z-huang/music#1409]
As there are many issues related to YouTube Music integration, I decided to move them into this issue. The features you’re asking contain as follows:

- Sync liked songs, playlists, albums, artists, and history
- Show YTM history
- Personalized suggestion
- Multiple channel support

It’s a huge work to implement this, from digging the API and considering data flow, to connecting the UI, and currently I don’t have a clear picture of how it should be like. Therefore, I need your suggestions, as ***detail*** as possible: How should local library and YTM library interact? Fully or partly?

**Comment by gmankab** [comment]
thank you for your very nice app ❤️

> How should local library and YTM library interact? Fully or partly?

partialy in first commits, but the goal full integration

> should we combine library songs and liked songs in order to be compatible with YTM library?

i think app should show 2 playlist, offline, and youtube music synced

offline one is always visible, and synced one is visible only if user logged in to youtube

i think here is a good roadmap for liked songs syncing commits:
- [ ] synced playlist visible in app
- [ ] make able playing music from synced playlist
- [ ] liked shongs should be added to synced playlist if user logged into acoount, if not then add to offline playlist
- [ ] allow to import csv with liked songs, [exported from youtube music](https://github.com/z-huang/InnerTune/issues/1409#issuecomment-2294895656) to any playlist (synced or offline one)
- [ ] allow to export songs from any playlist (offline and online one) to csv file, the file should be in same syntax like [official export from youtube music](https://github.com/z-huang/InnerTune/issues/1409#issuecomment-2294895656)
- [ ] sync all other playlist, besides liked songs
- [ ] sync all liked albums
- [ ] sync all liked artists

**Comment by gmankab** [comment]
how to use official export of liked songs from youtube music to csv:

- go to https://takeout.google.com/takeout/custom/youtube
- press all youtube data included
- deselect all
- select music-library-songs
- ok
- next step
- create export

**Comment by gmankab** [comment]
[gyawun](https://github.com/jhelumcorp/gyawun) and [simpmusic](https://github.com/maxrave-dev/SimpMusic) are good examles of apps with youtube music synchronisation

**Comment by Malopieds** [comment]
I think [OuterTune](https://github.com/DD3Boh/OuterTune) (a fork of InnerTune) also has it.

**Comment by gmankab** [comment]
> I think [OuterTune](https://github.com/DD3Boh/OuterTune) (a fork of InnerTune) also has it.

wow, first time seeing it, thank you a lot!

**Comment by z-huang** [comment]
> I think [OuterTune](https://github.com/DD3Boh/OuterTune) (a fork of InnerTune) also has it.

I tried, and it looks like songs, albums and artists are two-way synced. But playlist sync is not supported yet. I will figure out how the dev implement it.

**Comment by z-huang** [comment]
When a user turns on YTM sync, what should we do with the original local database? Merge, be replaced, or hide?

**Comment by mikooomich** [comment]
> But playlist sync is not supported yet.

What do you mean by this? I added a song to a playlist and it appears on my account.

If you're talking about creating a new playlist and adding to it, the syncing technically work, albeit some Ux issues (the create playlist dialog in the playlist screen isn't the updated version with the sync toggle, and playlists don't appear immediately right after creation)

Hi, I am one of the OuterTune devs

**Comment by z-huang** [comment]
@mikooomich I cannot add or delete songs for my YTM playlists.

**Comment by ruskcoder** [comment]
> I cannot add or delete songs for my YTM playlists.

It is working for me, I just signed in, and switch to the old interface layout, and all my ytm playlists show up in the library section. I usually use OuterTune to add/remove songs from my playlist.

**Comment by gmankab** [comment]
> When a user turns on YTM sync, what should we do with the original local database? Merge, be replaced, or hide?

good solution would be a "choose user" menu, that allow to choose betweeh offline user with old local database, and online user with new synced database

**Comment by Anonuser-code** [comment]
> When a user turns on YTM sync, what should we do with the original local database? Merge, be replaced, or hide?

Merge

**Comment by unseenlarks** [comment]
the 'account' section that leads to a user's YTM playlists, and the profile section found in settings that leads to the YTM home screen don't seem to be necessary, the first because those playlists could be on InnerTune's home screen after login, and the second because it shows the YTM UI which clashes as InnerTune's is much cleaner and easier to use.

if logged in, everything that belongs to the active YTM profile could be displayed on the main/home screen. if logged out, only local recommendations, albums, etc. would be there instead.

**Comment by Xisrr1** [comment]
There is another Music streaming app called 'Echo', which has YT music integration, and does a much better job at syncing everything and using recommendations from your account instead of locally.

**Comment by z-huang** [comment]
> There is another Music streaming app called 'Echo', which has YT music integration, and does a much better job at syncing everything and using recommendations from your account instead of locally.

Link please~

**Comment by gmankab** [comment]
https://github.com/brahmkshatriya/echo

**Comment by asifbagali** [comment]
java.lang.UnsupportedOperationException
	at k8.g.u(Unknown Source:5)
	at k8.g.e(SourceFile:10)
	at z8.t.u(SourceFile:58)
	at va.a.m(Unknown Source:8)
	at lb.n0.run(Unknown Source:100)
	at rb.a$a.run(SourceFile:82)

**Comment by NiceSapien** [comment]
Guys, is work being done on this? It would be very cool to have. Also, I had a question: Once this is implemented, would all my listening history from InnerTune appear in YouTube wrapped thingy at the end ot the year? This wrapped thingies is the reason I use spotify mods instead lol

**Comment by zettaexa** [comment]
> Guys, is work being done on this? It would be very cool to have. Also, I had a question: Once this is implemented, would all my listening history from InnerTune appear in YouTube wrapped thingy at the end ot the year? This wrapped thingies is the reason I use spotify mods instead lol

Yes since YouTube is getting data for Recaps through your watch history, which as you can see in the starting post is being worked on to be synced with YTM

**Comment by 1YuvrajS** [comment]
- [ ]  Make sure of the Music Video playback.

- [ ]  Screen Orientation, switching to Tablet mode for Android Head up Units.

**Comment by PhilippKosarev** [comment]
Personally I think it could be a good idea to have offline and online playlists; with the ability convert offline to online and vice versa. It's good for privacy since then it will only interact with the cloud when the user requests to.
Also just merging all playlists on login without asking the user could introduce a whole lot of issues and frustrated users (their asmr mommy roleplay playlist got leaked)

**Comment by hsp1020** [comment]
[simpmusic](https://github.com/maxrave-dev/SimpMusic) app updates all.

Sync liked songs, playlists, albums, artists, and history
Show YTM history
Personalized suggestion

**Comment by aneesh1122** [comment]
> [simpmusic](https://github.com/maxrave-dev/SimpMusic) app updates all.
> 
> Sync liked songs, playlists, albums, artists, and history
> Show YTM history
> Personalized suggestion

Add songs to a local playlist and add some songs. It won't show you the newly added songs. You have to restart the app to see those songs.

There are many bugs in that app.

**Comment by siggi1984** [comment]
https://github.com/Maloy-Android/Muzza

Here is another fork of Innertune that has YouTube Music integration. It's still in beta and some components are still missing. But it looks promising. Maybe you can check it out.

https://t.me/appmuzzaupdatesnews

**Comment by aneesh1122** [comment]
> https://github.com/Maloy-Android/Muzza
> 
> Here is another fork of Innertune that has YouTube Music integration. It's still in beta and some components are still missing. But it looks promising. Maybe you can check it out.
> 
> https://t.me/appmuzzaupdatesnews

I've been searching for a yt music player which can

1) sync likes and playlists with yt account
2) list all the cached songs
3) can set the cache limit
4) has mix playlists in the home page

**Comment by siggi1984** [comment]
> > https://github.com/Maloy-Android/Muzza
> > Here is another fork of Innertune that has YouTube Music integration. It's still in beta and some components are still missing. But it looks promising. Maybe you can check it out.
> > https://t.me/appmuzzaupdatesnews
> 
> I've been searching for a yt music player which can
> 
>     1. sync likes and playlists with yt account
> 
>     2. list all the cached songs
> 
>     3. can set the cache limit
> 
>     4. has mix playlists in the home page

Basically, you're looking for ViMusic with YTM integration. I haven't come across it yet.

**Comment by namename-123** [comment]
please add an option to view the YouTube ui directly, or better just to include what the site shows
I really want categories like "forgotten favorites" and "from the community" and I hope these are implemented in the future, thank you for the amazing app 🙏

**Comment by Maloy-Android** [comment]
Hello

**Comment by JonnyHaystack** [comment]
Personally I don't really *need* two-way sync as I would mostly manage playlists on desktop, and when I'm listening on my phone I just want to leave a playlist on shuffle.

It would be really nice if I could just make an unlisted playlist on youtube music and add it to my playlists on Innertune and have it always load the latest song list from that playlist. I imagine that would be a lot more straightforward than bidirectional syncing, and shouldn't even require logging in to YouTube (which is ideal to me).

Edit: Just figured out the above is already possible, though not very intuitive. If I send the YouTube playlist link (`youtube.com` **not** `music.youtube.com`) to my phone and Open in App -> Innertune, it opens properly in the app, and I can click the import button and thereafter long press and click Sync to update the local playlist.
